satpy/modifiers/spectral.py

```diff
@@ -107,7 +107,7 @@ def _get_sun_zenith_from_provided_data(projectables, optional_datasets):
             if sun_zenith_angle is None:
                 raise ImportError("Module pyorbital.astronomy needed to compute sun zenith angles.")
             _nir = projectables[0]
-            lons, lats = _nir.attrs["area"].get_lonlats(chunks=_nir.data.chunks)
+            lons, lats = _nir.attrs["area"].get_lonlats(chunks=_nir.data.chunks, dtype=_nir.dtype)
```

Comment:
I don't understand what the dtype of nir has to do with the type of longitudes and latitudes, och sun angles? I don't think it's the role of this function to decide on the type of the sun zenith...

Reply:
The dtype of sun angles is determined by the lon and lat dtypes. If the dtype of sun angles is `float64` it forces the 3.9 um reflectance to be `float64` no matter what the dtype of the input data is. NIR is the input data, so using its dtype is plain logical, and to save everything in between to be up- and then later downcast the best way to is to get the lonlats with the same dtype. Lons, lats and SZA are just intermediate data in the NIR reflectance determination, so it this saves a lot to use the intended type from the beginning without side-effects.

Reply:
Ok, I understand now that we need the sunz to be the same type as nir for later computations. But still, it's really hard to understand without context why the dtype is set to nir's, and could be confusing in the future.
Looking up a bit, it looks like this function is called from `_get_reflectance_as_dataarray` right? That function is at a higher abstraction level, and I think there it would make sense to decide the type of the sunz. So can we just pass down the dtype from there to `_get_sun_zenith_from_provided_data`?

Reply:
also, if we're modifying this function, it seems we are passing `projectables` here, while we just use the first element of the list, can we change that to just pass `nir` to the method?

Reply:
Ok, those make sens. Updated in https://github.com/pytroll/satpy/pull/2642/commits/4c7b3301d42bab04ed202e34c7f8fa4bfefee005